text/0000-javascript-module-api.md

```diff
@@ -0,0 +1,1058 @@
+
+For those who have already adopted modules via the `ember-cli-shims` package, we
+will provide a migration tool to rewrite shim modules into the final module API.
+The static nature of the import syntax makes this even easier and more reliable
+than migrating globals-based apps. The upgrade process should take no more than
+a few minutes (see [Migration](#migration)).
+
+# Detailed Design
+
+## Terminology
+
+* **Package** - a bundle of JavaScript addressable by npm and other package
+  managers, it may contain many modules (but has a default module, usually
+  called `index.js`).
+* **Scoped Package** - a namespaced package whose name starts with an `@`, like
+  `import Thing from "@scope/thing"`.
+* **Module** - a JavaScript file with at least one default export or named
+  export.
+* **Top-Level Module** - the module provided by importing a package directly,
+  like `import Component from "@ember/component"`.
+* **Nested Module** - a module provided at a path _inside_ a package, like
+  `import { addObserver } from "@ember/object/observers"`.
+
+Because our goal is to eliminate the `Ember` global object, any public classes,
+functions or properties that currently exist on the global need an equivalent
+module that can be imported.
+
+Given how fundamental modules are to the development process, how we organize
+and name them impacts new learners and seasoned veterans alike. Thus we must try
+to find a balance between predictability for new and intermediate users, and
+terseness for experienced developers with large apps.
+
+There is another goal at play: we would like to help dispel the misconception
+that Ember is a monolithic framework. Ideally, module names help us tell a story
+about Ember's layered features. Rather than inheriting the entire framework at
+once, you can pull in just the pieces you need.
+
+For that reason, package names should assist the developer in understanding what
+capabilities are added by bringing in that new package. We should pick
+meaningful names, not let our public API be a by-product of how Ember's
+internals are organized.
+
+A full table of proposed mappings from global to module is available in
+[Addendum 1 - Table of Module Names and Exports by
+Global](#addendum-1---table-of-module-names-and-exports-by-global) and [Addendum
+2 - Table of Module Names and Exports by
+Package](#addendum-2---table-of-module-names-and-exports-by-package). Because
+there is some implicit functionality that you get when loading Ember that is not
+encapsulated in a global property (for example, automatically adding prototype
+extensions), there is also [Addendum 3 - Table of Modules with Side
+Effects](#addendum-3---table-of-modules-with-side-effects).
+
+Before diving in to these tables, however, it may be helpful to understand some
+of the thinking that guided this proposal.
+
+### Use Scoped Packages
+
+Last year, [npm introduced support for scoped packages][scoped-packages]. Scopes
```

Comment:
Hm, I've had success with non-private scoped packages. If they don't play nicely, we should definitely reach out and potentially file a bug!